ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Wallets.hs

```diff
@@ -67,43 +67,54 @@ mkHasPP pp = HasNonemptyPassphrase $ pp /= emptyPassphrase
-             -> IO (Either CreateWalletError HdRoot)
+createHdWallet
+    :: PassiveWallet
+    -> EncryptedSecretKey
+    -- ^ Wallet's encrypted secret key. Should be generated outside.
+    -> HasNonemptyPassphrase
+    -- ^ Whether the newly created wallet has a non-empty passphrase.
+    -> AssuranceLevel
+    -- ^ The 'AssuranceLevel' for this wallet, namely after how many
+    -- blocks each transaction is considered 'adopted'. This translates
+    -- in the frontend with a different threshold for the confirmation
+    -- range (@low@, @medium@, @high@).
+    -> WalletName
+    -- ^ The name for this wallet.
+    -> UtxoByAccount
+    -- ^ Initial utxo for the new wallet.
+    -> IO (Either CreateWalletError HdRoot)
 createHdWallet pw esk hasNonemptyPassphrase assuranceLevel walletName utxoByAccount = do
+    -- Note: upstream implementation inserted a new HdRoot into acid-state
+    -- first, and then added the new key into the keystore. We do
+    -- this in the reverse order because we want to maintain the
+    -- invariant where having an HdRoot in the acid-state guarantees
+    -- that the corresponding key is present in the keystore.
+    --
     -- STEP 1: Insert the 'EncryptedSecretKey' into the 'Keystore'
-    let newRootId = HD.eskToHdRootId esk
-        walletId = WalletIdHdSeq newRootId
-    -- This may throw an IO exception.
-    Keystore.insert walletId esk (pw ^. walletKeystore)
-    -- STEP 2: Atomically generate the wallet and the initial internal structure in
-    -- an acid-state transaction.
-    res <- createWalletHdSeq
-        pw
-        hasNonemptyPassphrase
-        walletName
-        assuranceLevel
-        esk
-        utxoByAccount
-    case res of
-        Left e -> do
-            Keystore.delete walletId (pw ^. walletKeystore)
-            return . Left $ CreateWalletFailed e
-        Right hdRoot -> do
-            return (Right hdRoot)
-
+    -- This may throw an IO exception, which we rewrap into Either
+    -- in order to follow the general pattern of imported code.
+    let hdrId = HD.eskToHdRootId esk
+        walletId = WalletIdHdSeq hdrId
+    res1 <- try $ Keystore.insert walletId esk (pw ^. walletKeystore)
+    case res1 of
+        Left Keystore.DuplicatedWalletKey ->
+            pure $ Left $ CreateWalletFailed (HD.CreateHdRootExists hdrId)
+        Right () -> do
+            -- STEP 2: Atomically generate the wallet and the initial internal structure in
+            -- an acid-state transaction.
+            res2 <- createWalletHdSeq
+                pw
+                hasNonemptyPassphrase
+                walletName
+                assuranceLevel
+                esk
+                utxoByAccount
+            case res2 of
+                Left e -> do
+                    liftIO $ Keystore.delete walletId (pw ^. walletKeystore)
+                    pure $ Left $ CreateWalletFailed e
+                Right hdRoot -> do
+                    pure $ Right hdRoot
```

Comment:
Is the only change here that we catch `DuplicatedWalletKey` and turn it into `CreateWalletFailed`?